romdisk/date.asm

```diff
@@ -49,8 +92,117 @@ date_main:
+        S_WRITE1(DEV_STDOUT)
+        ret
+
+        ; Read ACSII date and fill out date part of DATE_STRUCT.
+        ; The input format will be as follows:
+        ;   YYYY-MM-DD
+        ; Does not check the '-' chars, they can be anything. Does not
+        ; validate number ranges.
+        ; Parameters:
+        ;       HL - Pointer to the date structure, of size DATE_STRUCT_SIZE
+        ;       DE - String source.
+        ; Returns:
+        ;       A - nonzero if bad format detected. 
+        ;       If A is zero the date parts of the input struct have been filled out.
+        ;       DE - DE + 10
+        ; Modifies: HL, BC
+        PUBLIC ascii_to_date
+ascii_to_date:
+        xor a
+        ld b, '-'
+        call read_bcd_pair
+        call read_bcd_pair
+        call check_char
+        call read_bcd_pair
+        call check_char
+        call read_bcd_pair
+        ; TODO calculate and fill out day-of-week
+        ret
+
+        ; Read ACSII time and fill out time part of of DATE_STRUCT.
+        ; The input format will be as follows:
+        ;   HH:MM:SS
+        ; Does not check the ':' chars, they can be anything. Does not
+        ; validate number ranges.
+        ; Parameters:
+        ;       HL - Pointer to the time part of date structure
+        ;       DE - String source.
+        ; Returns:
+        ;       A  - nonzero if bad format detected
+        ;       The time parts of the input date struct have been filled out.
+        ; Modifies: BC
+        PUBLIC ascii_to_time
+ascii_to_time:
+        xor a
+        ld b, ':'
+        call read_bcd_pair
+        call check_char
+        call read_bcd_pair
+        call check_char
+        call read_bcd_pair
+        ret
+
+        ; Convert an decimal ascii pair of bytes to a BCD byte.
+        ; Parameters
+        ;       DE - Pointer to first (high) ascii byte
+        ;       HL - Pointer to result byte to update
+        ; Returns:
+        ;       (HL) - 2 BCD nibbles, big endian
+        ;       HL   - HL + 1
+        ;       DE   - DE + 2
+        ;       A    - unmodified if both input chars are digit,
+        ;              nonzero otherwise
+read_bcd_pair:
+        ld c, a ; we use C as accumulated error condition
+        ld a, (de)
+        cp '9' + 1
+        jr c, _ok1
+        ld c, 1
+_ok1:   sub '0'
+        jr nc, _ok2
+        ld c,1
+_ok2:   ld (hl),a
+        inc de ; next digit
+        ld a, (de)
+        cp '9' + 1
+        jr c, _ok3
+        ld c,1
+_ok3:   sub '0'
+        jr nc, _ok4
+        ld c,1
+_ok4:   rld
+        inc de
+        inc hl
+        ld a, c ; return original value or 1 if erro detected
+        ret
+
+check_char:
+        ; Check that char matches expected one.
+        ; Parameters
+        ;       B  - expected char
+        ;       DE - Pointer to char to check
+        ; Returns:
+        ;       A  - nonzero on mismatch, unmodified on match
+        ;       DE - DE + 1
+        ; Modifies: C
+        ld c, a
```

Comment:
If we follow the nz-flag-on-error approach, this routine can be converted to:
```
check_char:
    ld a, (de)
    sub b
    ret
```

Reply:
Did this